File: 01-SRP/cpp/refactored.cpp
#include <iostream>
#include <string>

struct Order {
    int id;
    std::string item;
    double price;
};


class OrderRules {
public:
    void validate(const Order& order) {
        std::cout << "Validating order\n";
    }

    double calculatePrice(const Order& order) {
        std::cout << "Calculating price\n";
        return 1000.0;
    }
};


class OrderRepository {
public:
    void save(const Order& order) {
        std::cout << "Saving order to database\n";
    }
};


class EmailService {
public:
    void sendConfirmation(const Order& order) {
        std::cout << "Sending confirmation email\n";
    }
};


class AuditLogger {
public:
    void logEvent(const Order& order) {
        std::cout << "Logging order event\n";
    }
};


class OrderService {
private:
    OrderRules& rules;
    OrderRepository& repository;
    EmailService& emailService;
    AuditLogger& logger;

public:
    OrderService(
        OrderRules& rules,
        OrderRepository& repository,
        EmailService& emailService,
        AuditLogger& logger
    )
        : rules(rules),
          repository(repository),
          emailService(emailService),
          logger(logger) {}

    void createOrder(Order& order) {
        rules.validate(order);

        order.price = rules.calculatePrice(order);

        repository.save(order);
        emailService.sendConfirmation(order);
        logger.logEvent(order);
    }
};


int main() {
    Order order{101, "Laptop", 0.0};

    OrderRules rules;
    OrderRepository repository;
    EmailService emailService;
    AuditLogger logger;

    OrderService service(
        rules,
        repository,
        emailService,
        logger
    );

    service.createOrder(order);

    return 0;
}

File: 01-SRP/python/refactored.py
class OrderRules:
    def validate(self, order):
        print("Validating order")

    def calculate_price(self, order):
        print("Calculating price")
        return 1000


class OrderRepository:
    def save(self, order):
        print("Saving order to database")


class EmailService:
    def send_confirmation(self, order):
        print("Sending confirmation email")


class AuditLogger:
    def log_event(self, order):
        print("Logging order event")


class OrderService:
    def __init__(self, rules, repository, email_service, logger):
        self.rules = rules
        self.repository = repository
        self.email_service = email_service
        self.logger = logger

    def create_order(self, order):
        self.rules.validate(order)

        order["price"] = self.rules.calculate_price(order)

        self.repository.save(order)
        self.email_service.send_confirmation(order)
        self.logger.log_event(order)


if __name__ == "__main__":
    order = {
        "id": 101,
        "items": ["Laptop", "Mouse"]
    }

    rules = OrderRules()
    repository = OrderRepository()
    email_service = EmailService()
    logger = AuditLogger()

    service = OrderService(
        rules,
        repository,
        email_service,
        logger
    )

    service.create_order(order)